lectures/newtons_method.md

```diff
@@ -0,0 +1,469 @@
+
+Assuming we have such market for two complementary goods where demand depends on the price of both components. 
+
+We label them $\text{good}_0$ and $\text{ good}_1$, with price vector $p = (p_0, p_1)$.
+
+Then the supply of good $i$ at price $p$ is,
+
+$$ 
+q^s_i (p) = b_i \sqrt{p_i} 
+$$
+
+and the demand of good $i$ at price $p$ is,
+
+$$ 
+q^d_i (p) = \text{exp}(-(a_{i0} p_0 + a_{i1} p_1)) + c_i
+$$
+
+Here $a_{ij}$, $b_i$ and $c_i$ are parameters for $n \times n$ square matrix $A$ and $n \times 1$ parameter vectors $b$ and $c$.
+
+The excess demand function is,
+
+$$
+e_i(p) = q_i^d(p) - q_i^s(p), \quad i = 0, 1
+$$
+
+An equilibrium price vector $p^*$ satisfies $e_i(p^*) = 0$.
+
+We set
+
+$$
+A = \begin{pmatrix}
+            a_{00} & a_{01} \\
+            a_{10} & a_{11}
+        \end{pmatrix},
+            \qquad 
+    b = \begin{pmatrix}
+            b_0 \\
+            b_1
+        \end{pmatrix}
+    \qquad \text{and} \qquad
+    c = \begin{pmatrix}
+            c_0 \\
+            c_1
+        \end{pmatrix}
+$$
+
+for this particular question.
+
+
+### The Multivariable Market Equilibrium
+
+We can now easily get the multivariable version of the problem above.
+
+The supply function remain unchanged,
+
+$$ 
+q^s (\bf{p}) =\bf{b} \bf\sqrt{p} 
+$$
+
+The demand function is,
+
+$$ 
+q^d (\bf{p}) = \text{exp}(- \bf{A} \cdot \bf{p}) + \bf{c}
+$$
+
+Our new excess demand function is,
+
+$$
+e(\bf{p}) = \text{exp}(- \bf{A} \cdot \bf{p}) + \bf{c} - \bf{b} \bf\sqrt{p} 
+$$
+
+The function below calculates the excess demand for the given parameters
+
+```{code-cell} python3
+def e(p, A, b, c):
+    return jnp.exp(- A @ p) + c - b * jnp.sqrt(p)
+```
+
+
+## Using Newton's Method
+
+Now let's use the multivariate version of Newton's method to compute the equilibrium price
+
+```{math}
+:label: multi-newton
+p_{n+1} = p_n - J_e(p_n)^{-1} e(p_n)
+```
+
+Here $J_e(p_n)$ is the Jacobian of $e$ evaluated at $p_n$.
+
+The iteration starts from some initial guess of the price vector $p_0$. 
+
+Here, instead of coding Jacobian by hand, We use the `jax.jacobian()` function to auto-differentiate and calculate the Jacobian.
+
+With only slight modification, we can generalize [our previous attempt](https://python.quantecon.org/newton_method.html#first_newton_attempt) to multi-dimensional problems
+
+```{code-cell} python3
+def newton(f, x_0, tol=1e-5, max_iter=10):
+    x = x_0
+    q = jax.jit(lambda x: x - jnp.linalg.solve(jax.jacobian(f)(x), f(x)))
+    error = tol + 1
+    n = 0
+    while error > tol:
+        n+=1
+        if(n > max_iter):
+            raise Exception('Max iteration reached without convergence')
+        y = q(x)
+        if(any(jnp.isnan(y))):
+            raise Exception('Solution not found with NaN generated')
+        error = jnp.linalg.norm(x - y)
+        x = y
+        print(f'iteration {n}, error = {error:.5f}')
+    print('\n' + f'Result = {x} \n')
+    return x
+```
+
```

Comment:
It might be good to call this function on a test example.

Then if the code is broken by some future change to JAX, we will get a build failure that alerts us

Reply:
Thanks @jstac,this function is called below at line 199 and I think we don't need to write another test sample because it is already called.

Reply:
You are right @HengchengZhang , my bad.